Index: src/org/aspectj/util/ConfigParser.java
===================================================================
RCS file: /home/technology/org.aspectj/modules/util/src/org/aspectj/util/ConfigParser.java,v
retrieving revision 1.6
diff -u -r1.6 ConfigParser.java
--- src/org/aspectj/util/ConfigParser.java	5 Aug 2004 17:31:56 -0000	1.6
+++ src/org/aspectj/util/ConfigParser.java	2 Sep 2005 07:40:04 -0000
@@ -21,11 +21,14 @@
 public class ConfigParser {
     Location location;
     protected File relativeDirectory = null;
-    protected List files = new LinkedList();
+    protected List files = new ArrayList();
+    protected Map userFileMap = new HashMap(); // from File -> File, keys are canonical files, values are the 'user' path
     private boolean fileParsed = false;
     protected static String CONFIG_MSG = "build config error: ";  
     
     public List getFiles() { return files; }
+    
+    public Map getUserFileNameMap() { return userFileMap; }
 
     public void parseCommandLine(String[] argsArray) throws ParseException {
         location = new CommandLineLocation();
@@ -221,7 +224,9 @@
           ret = new File(dir, name);
         }
         try {
-        	ret = ret.getCanonicalFile();
+        	File canonicalFile = ret.getCanonicalFile();
+        	userFileMap.put(canonicalFile,new File(name));
+        	ret = canonicalFile;
         } catch (IOException ioEx) {
         	// proceed without canonicalization
         	// so nothing to do here
